### ISSUE TITLE
Wrong permission used for users with capabilities and document_read_uses_read ignored

### ISSUE THREAD

**Issue by kwisatz** [wp-document-revisions/wp-document-revisions#256]
### Describe the bug

The capabilities assigned to user roles include `read_documents`. However, in `serve_document_auth`, the capability that is checked is `read_document` (without the `s`).

`document_read_uses_read` is not observed in this context.

### Steps to reproduce the behavior

Attempt to navigate to a document as e.g. administrator.

### Expected behavior

The document should be served to authenticated users.

### Additional context

Theoretically, `map_meta_cap` would change the permission name to `read_documents`, however, this only happens if `apply_filters( 'document_read_uses_read', true )` returns `false`:

```php
	/**
	 * Determines if read requires read_document instead of read.
	 *
	 * @since 3.3
	 */
	public function use_read_capability() {
		// user requires read_document and not just read to read document.
		/**
		 * Filters the users capacities to require read (or read_document) capability.
		 *
		 * @since 3.3
		 *
		 * @param boolean true  default action to capability read for documents (not read_document).
		 */
		if ( ! apply_filters( 'document_read_uses_read', true ) ) {
			// invoke logic.
			add_filter( 'map_meta_cap', array( &$this, 'map_meta_cap' ), 10, 4 );
			add_filter( 'user_has_cap', array( &$this, 'user_has_cap' ), 10, 4 );
			if ( ! current_user_can( 'read_documents' ) ) {
				// user does not have read_documents capability, so any need to be filtered out of results.
				add_filter( 'posts_results', array( &$this, 'posts_results' ), 10, 2 );
			}
		}

		// filter the queries.
		add_action( 'pre_get_posts', array( &$this, 'retrieve_documents' ) );
	}
```

If the default behavior is kept, then `add_filter('map_meta_cap', array( &$this, 'map_meta_cap' ), 10, 4 );` is never called and the plug-in's `map_meta_cap` method will never be called from WP's capabilities check.

This call 

```php
printf('<!-- DEBUG %s-->', json_encode(apply_filters( 'map_meta_cap', ['read_document'], 'read_document', 20, [$post->ID] )));
```

will thus just return:

```html
<!-- DEBUG ["read_document"]--><!DOCTYPE html>…
```

In `serve_document_auth` we can see that no further check is made for the status of `document_read_uses_read` once we passed the first two checks. The last conditional here which applies to users with the administrator role will unconditionally check for `read_document` which does not exist and is never converted into `read_documents` by the `map_meta_cap` filter.

```php
	public function serve_document_auth( $default, $post, $version ) {
		$user     = wp_get_current_user();
		$ret_null = ( 0 === $user->ID && ! apply_filters( 'document_read_uses_read', true ) );
		// public file, not a revision, no need to go any further
		// note: non-authenticated users only have the "read" cap, so can't auth via read_document.
		if ( ! $version && 'publish' === $post->post_status ) {
			if ( 0 === $user->ID && apply_filters( 'document_read_uses_read', true ) ) {
				// Not logged on. But only default read capability.
				return $default;
			}
		}

		// need to check access.
		// attempting to access a revision.
		if ( $version && ! current_user_can( 'read_document_revisions' ) ) {
			return ( $ret_null ? null : false );
		}

		// specific document cap check.
		if ( ! current_user_can( 'read_document', $post->ID ) ) {
			return ( $ret_null ? null : false );
		}

		return $default;
	}
```

**Comment by NeilWJames** [comment]
@kwisatz 
Please can I make sure that I understand the actual problem that you're having. Is it that:
a) you have set the filter `document_read_uses_read` to return false
b) a user with the read_documents capability is NOT able to read a document.
Am I correct please?

Note that the function map_meta_cap is always called - because it is in the post_type definition.

I did find it quite complex to figure out how this all works when reviewing and creating this code (and the corresponding taxonomy permissions stuff in the Cookbook). Some findings were
1. Out of the box, everyone (logged-in or not) has read capability so all documents can be read. Thus read_documents permissions is thus actually irrelevant normally.
2. You use current_user_can with its plural form to see if you have the generic capability ( and it is that is managedwith the underlying has_cap. So you need edit_documents capability to create a post.
3. However when you want to test whether the user has the capability for a specific post, then you need to use the singular form.

There has been some correspondence on trac as to whether "doing_it_wrong" should be called in such cases.

Thus, for me, your first sentence is a correct statement of the required behaviour.

So if you look at what the functions are doing (and only in the case that we don't want to use the "read" capability), it is:
1. In the filter map_meta_cap (part of the function map_meta_cap), will change the required "read_post" into "read_document" if the post is a document. Standard WP processing has already transformed the original "read_document" into "read_post" - the base capability.
2.  In the filter user_has_cap (used to return the user capabilities to check the required capabilities), we bail in the case of an administrator (actually not necessary, but being safe), then for a document to remove the "read" capability. And if the user has "read_documents" capability, then they will be able to access the file.

Finally the "post_results" filter is there to stop any WP_Query returning documents if you don't have "read_documents" capability.

Regards,
Neil James

**Comment by kwisatz** [comment]
> Please can I make sure that I understand the actual problem that you're having. Is it that:
> a) you have set the filter document_read_uses_read to return false

No, it's at the default of `true`. Which is also why this never runs: `add_filter( 'map_meta_cap', array( &$this, 'map_meta_cap' ), 10, 4 );`.

> b) a user with the read_documents capability is NOT able to read a document.

Correct. An anonymous user was, an administrator in our case was not.

> Note that the function map_meta_cap is always called - because it is in the post_type definition.

I'm positive it was not, I had code and an `exit;` statement added to the function, I'm very sure it is never called.
I don't know when or where the function gets called when you add it to a post_type registration, I'd have to trace that or read the code. But from `'map_meta_cap'         => true,` alone, I don't see how this would ever invoke your method.

The place where it would have to be called in order to transform the check `current_user_can('read_document', $post->ID)` into `current_user_can('read_documents')` would be in the context of `current_user_can() --> \WP_User->has_cap()`.

> Thus, for me, your first sentence is a correct statement of the required behaviour.

OK, I didn't check the cookbook on this. But the only place in the code where I could find that this transformation from singular to plural could be taking place would be `map_meta_cap`, which again, never runs in our case. 

I'll be happy to retrace all of this again, yet not on the same site, because I eventually broke it yesterday and without access to the filesystem had to ask support to fix it which took 12 hours…
But, given that `document_read_uses_read` is not set to return `false` in our case, maybe you can run through your code again and see if it still works out?

**Comment by NeilWJames** [comment]
@kwisatz ,
Thanks for your reply. I may have confused things with the term map_meta_cap. The same term is used in a number of places.

When the post_type is registered with `'map_meta_cap' => true,`, it is saying that there will be capabilities of `edit_documents` instead of `edit_posts`  etc being looked up by standard WP functionality. This is managed by the function map_meta_cap and is contained in /includes/wp-capabilities.php. It is this function that is always called that I referred to earlier. Sorry for any confusion.

It contains a filter called map_meta_cap which may link to the function `WP_Document_Revisions::map_meta_cap`.

We can agree that this function is never called if `apply_filters( 'document_read_uses_read', true )` returns `true`.

It would be an error if it were called. Why? Because the job of it and user_has cap is to dynamically remove `read` cabability from the user and to test against the `read_documents` capability when accessing documents.

I have experienced your problem. 

Please check that ALL roles associated with your administrator user have the required access and none have it set as Deny. 

I had the experience that my administrator user had a secondary role where edit_documents capability was set to false. So despite the main role having edit_documents set to true, I was unable to see All Documents in the Admin menu.

Possibly this is the cause of your issue. You may have previously deleted an xxx_documents capability from a secondary role. On plugin upgrade, the plugin is essentially deactivated, updated and then reactivated. The last process will create the "missing" capabilities with their default values.

FWIW In the preamable to WP map_meta_cap contains this comment:
` * Maps a capability to the primitive capabilities required of the given user to
 * satisfy the capability being checked.
 *
 * This function also accepts an ID of an object to map against if the capability is a meta capability. Meta
 * capabilities such as `edit_post` and 'edit_user' are capabilities used by this function to map to primitive
 * capabilities that a user or role requires, such as 'edit_posts' and 'edit_others_posts'.`

It is this function that, as it says, maps the meta capabililities (`edit_document`) into the primitive capability (`edit_others_documents`). That is why it is not in the plugin.

Hope this helps,
Neil James

**Comment by NeilWJames** [comment]
@kwisatz,
I have been able to replicate your issue in a specific case. That is, where the filter `document_read_uses_read` returns true and a signed-on user cannot read the document.

This is in the case the I removed read access from the user and would rely on the read_documents capabilty to read the document. This does not appear to work.

That is, I've told it that it uses read access to access the document and then I haven;t given the user the access. At one level, it is correct that the user should not have access.

There does appear to be an assumpion in the WP function that you do always have read access and the mapping of read_post to read_documents does not appear to happen, so signed-on users have no access. (Note code assumes that non signed-on users have read access and there is no need to test them for accessing public documents.)

Whilst always calling `add_filter( 'map_meta_cap', array( &$this, 'map_meta_cap' ), 10, 4 );` could give you relief in this specific scenario, it is unclear to me without a much deeper understanding of the operation of the standard routine whether it is a general fix.

Hope this give some ideas,
Neil James

**Comment by NeilWJames** [comment]
This note is to document the result of the deeper investigations carried out as a result of this issue - and to document the simplification of the plugin code as a result to be released in due course.

I was unable to find an issue in this area of the code.

Firstly it shoud be noted in what follows there are 3 base meta capabilities `read_post`, `edit_post` and `delete_post`. In what follows, I will only refer to `read_post` - this is for brevity, but it applies to the other two as well.

`read_post` is a meta capability because it does not exist as a capability, but is used as part of the mapping process within the map_meta_cap function within /wp_includes/capabilities.php.

When defining the post_type, it uses these parameters:
`			'capability_type'      => array( 'document', 'documents' ),
			'map_meta_cap'         => true,`

The effect of this is that that the meta capability `read_document` is deifned that will map to `read_post` and other real capabilities are created, including `read_documents`.

When testing for capabilities with current_user_can, we should use the real capabilility name (plural) when wanting to test wither they have the generic capability, but the meta capability name (singular) when testing against a specific post, This is because when there is a specific post, we need to map down to `read` or `read_private_documents`. Note that I state `read` because that is the standard WP functionality. I will come back to this later.

So when we ask for `read_document` capability, the routine will map this to `read_post` (because map_meta_cap has been set to true) and it will go through the generic processing. and come up with either `read` or `read_private"documents` or `do_not_allow` (for an unpublished document written by another).

So the result of the standard processing with these parameters is that `read` capability is required to read a document - and if we we want that `document_read_uses_read` is set to false, we don't want this, we want `read_documents` capability to be checked.

And therefore the plugin routines `map_metability is_cap` and `user_has_cap` exist to ensure the correct capability is checked.

As a result of this investigation, I have found that by adding the parameter to the post_type registration:
`			'capabilities'         => array(
				'read' => ( apply_filters( 'document_read_uses_read', true ) ? 'read' : 'read_documents' ),
				),`

In this way we can get the `read` capability to be mapped to the correct permission - and these functions can be removed as the correct permission is being required.

Sorry that this sheds no further light on the original problem raised.

Neil James

**Comment by NeilWJames** [comment]
Addressed by #265
